{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3f839d60",
   "metadata": {},
   "source": [
    "### Overview\n",
    "\n",
    "- Showcases steps involved in the initial \"patchification\" process.\n",
    "- Pre-patchifying the dataset instead of doing it on-the-fly considerably reduces runtime by avoiding expensive merge/table joins with the reference genome loci.\n",
    "\n",
    "The 16 example GTEx samples (pre-reduced to only contain chr21) will undergo 2 transformation:\n",
    "1. Simulated to be RRBS-like missing using the \"rrbs_sim\" noise regime. \n",
    "    - This is to showcase model inference with the supplied pre-trained model in checkpoints/\n",
    "2. Simulated to be MCAR at a 0.9 missing ratio.\n",
    "    - This is to supply example data for training a new model. \n",
    "\n",
    "* NOTE: This code will create process and store additional data in the CWD!\n",
    "* NOTE: process_dataset assumes reference genome pre-computed metadata is in CWD/data/metadata. If not, please change path in ```aruna.process_dataset```."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bea53ced",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/js228/ARUNA\n"
     ]
    }
   ],
   "source": [
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6619e4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from aruna.process_dataset import get_cc_gt, get_cc_noisy\n",
    "from aruna.process_dataset import get_pc_gt, get_pc_noisy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7f9ba6d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "CWD = os.getcwd()\n",
    "DATA_DIR = os.path.join(CWD, \"data\", \"gtex_subset\") # dir to all sample BED-like files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333c8c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"gtex\"\n",
    "chrom = \"chr21\"\n",
    "num_cpg = 128\n",
    "nr = \"mcar_90\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a7751c8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Curent chromosome:  chr21\n",
      "Looking for Patchified Noise-Simulated FractionalMethylation and Coverage files...\n",
      "One or both files not found!\n",
      "Processing for gtex - chr21 - mcar_90 and 128 CpGs/Patch...\n",
      "PatchID-CpG reference maps complete!\n",
      "Looking for Noise-Simulated FractionalMethylation and Coverage files...\n",
      "One or both files not found. Creating...\n",
      "Looking for Ground-Truth FractionalMethylation and Coverage files...\n",
      "One or both files not found.\n",
      "Processing for gtex and chr21...\n",
      "cpgMerged.CpG_report.merged_CpG_evidence.cov files for 16 samples found!\n",
      "Creating FM and COV files with 8 workers...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:01<00:00, 13.10it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CpG Observation rates for gtex - chr21 over ALL samples are: \n",
      "Mean:  84.549\n",
      "Std:  0.931\n",
      "For gtex - chr21, saving Fractional Methylation Data for all samples to: /home/js228/ARUNA/data/gtex/chrom_centric/true/FractionalMethylation/chr21.fm...\n",
      "For gtex - chr21, saving Read Coverage Data for all samples to: /home/js228/ARUNA/data/gtex/chrom_centric/true/ReadDepth/chr21.cov...\n",
      "All files saved to disk!\n",
      "Processing for gtex - chr21 - mcar_90...\n",
      "Creating Noise Simulation files...\n",
      "Computing metrics for sanity checks...\n",
      "Avg: 15.451 Std: 0.931 of Originally Missing CpG rates.\n",
      "\n",
      "Avg: 89.994 Std: 0.046 of Simulated Missing CpG rates.\n",
      "\n",
      "Avg: 91.546 Std: 0.106 of After-Simulation Missing CpG rates.\n",
      "\n",
      "For gtex - chr21, saving Fractional Methylation Data for all samples to: /home/js228/ARUNA/data/gtex/chrom_centric/mcar_90/FractionalMethylation/chr21.mask.fm...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/js228/ARUNA/aruna/noise_simulators.py:22: UserWarning: NR seed not specified! Defaulting to seed = 0.\n",
      "  warnings.warn(\"NR seed not specified! Defaulting to seed = 0.\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved!\n",
      "For gtex - chr21, saving Simulated Mask Data for all samples to: /home/js228/ARUNA/data/gtex/chrom_centric/mcar_90/SimulatedMask/chr21.mask...\n",
      "Saved!\n",
      "\n",
      "All files saved to disk!\n",
      "Creating mappings such as sample: [patch_id: patch_betas] with 8 workers...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:08<00:00,  1.85it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patchification complete! Saving FM Data for all samples to: /home/js228/ARUNA/data/gtex/patch_centric/numCpg128/mcar_90/FractionalMethylation/chr21_patches.mask.fm.pkl...\n",
      "Saved!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:08<00:00,  1.94it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patchification complete! Saving SimMask Data for all samples to: /home/js228/ARUNA/data/gtex/patch_centric/numCpg128/mcar_90/SimulatedMask/chr21_patches.mask.pkl ...\n",
      "Saved!\n",
      "All files saved to disk!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/js228/ARUNA/aruna/process_dataset.py:210: UserWarning: Since sample subset is not provided, returning entire dataset!\n",
      "  warnings.warn(\"Since sample subset is not provided, returning entire dataset!\")\n"
     ]
    }
   ],
   "source": [
    "pc_nrfm_map, pc_nrmask_map = get_pc_noisy(dataset = dataset, chrom = chrom, \n",
    "                                          num_cpg = num_cpg, \n",
    "                                          nr = nr, data_dir = DATA_DIR)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "XAI",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
