{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0753b418",
   "metadata": {},
   "source": [
    "### Overview\n",
    "\n",
    "This notebook demonstrates the ARUNA inference pipeline for upscaling sparse, sequencing-based DNA methylation data to genome-wide resolution using a pretrained model.\n",
    "\n",
    "Inference assumes that input data (either simulated RRBS-like data or real RRBS data) have already been converted into the **patch-centric representation** expected by the model. In particular, input patches must match the patch size used during training (here, 128 CpGs per patch). For details see data_prep.ipynb.\n",
    "\n",
    "While a pretrained ARUNA model can, in principle, be applied across a wide range of sparsity levels, the provided model and hyperparameters were specifically tuned for extreme sparsity regimes, including:\n",
    "- MCAR missingness ≥ 90%\n",
    "- RRBS-like missing-not-at-random patterns\n",
    "\n",
    "---\n",
    "\n",
    "### Requirements\n",
    "\n",
    "> **Note:** This notebook assumes that all preprocessing steps in `data_prep.ipynb` have already been completed.\n",
    "\n",
    "The following artifacts must be available:\n",
    "\n",
    "- **Pretrained ARUNA model**  \n",
    "  A PyTorch checkpoint (`.pth`) located under:\n",
    "`ARUNA/checkpoints/`\n",
    "\n",
    "- **Patchified sparse methylation data**  \n",
    "Patch-centric data generated by `data_prep.ipynb`, stored under:\n",
    "`ARUNA/data/<dataset>/`\n",
    "\n",
    "---\n",
    "\n",
    "### Outputs\n",
    "\n",
    "Running this notebook will produce:\n",
    "- Chromosome-wide predicted methylomes and associated evaluation masks written to: `ARUNA/results/`\n",
    "- Example output format:\n",
    "```text\n",
    "results/\n",
    "├── eval_mask\n",
    "│   └── chr21.csv\n",
    "└── pred_betas\n",
    "    └── chr21.csv\n",
    "```\n",
    "- eval_mask: Boolean masks for every chr CpG. True = CpG was observed in (known or matched) ground truth and either simulated to be missing or missing in the corresponding RRBS.\n",
    "- pred_betas: The predicted beta values at each chr CpG.\n",
    "\n",
    "\n",
    "These outputs can be directly used for downstream evaluation, visualization, or comparison with baseline imputation methods.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b8a71382",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/js228/ARUNA\n"
     ]
    }
   ],
   "source": [
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "48bad8d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "from aruna.process_dataset import get_cc_gt\n",
    "from scripts.inference import get_cpgmask, save_aruna_preds, run_mslice_inference\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "57c04451",
   "metadata": {},
   "outputs": [],
   "source": [
    "CWD = os.getcwd()\n",
    "model_fpath = os.path.join(CWD, \"checkpoints\", \"trained_model.pth\")\n",
    "config_fpath = os.path.join(CWD, \"configs\", \"example_config.yaml\")\n",
    "res_dir = os.path.join(CWD, \"results\") # save path for preds and evalMask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c68ea00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for Ground-Truth FractionalMethylation and Coverage files...\n",
      "All files found!\n",
      "FM at: /home/js228/ARUNA/data/gtex/chrom_centric/true/FractionalMethylation/chr21.fm\n",
      "COV at: /home/js228/ARUNA/data/gtex/chrom_centric/true/ReadDepth/chr21.cov\n",
      "Looking for Ground-Truth FractionalMethylation and Coverage files...\n",
      "All files found!\n",
      "FM at: /home/js228/ARUNA/data/gtex/chrom_centric/true/FractionalMethylation/chr21.fm\n",
      "COV at: /home/js228/ARUNA/data/gtex/chrom_centric/true/ReadDepth/chr21.cov\n"
     ]
    }
   ],
   "source": [
    "test_data = \"gtex\"\n",
    "chrom = \"chr21\"\n",
    "test_regime = \"rrbs_sim\"\n",
    "save_path = Path(CWD) / \"results\"\n",
    "save_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "cpgMask_map = get_cpgmask([chrom], test_data) # cpgs missing in ground truth\n",
    "\n",
    "cc_gt_df, _ = get_cc_gt(test_data, chrom)\n",
    "canonical_index = cc_gt_df.index # canonical set of hg38 CpGs\n",
    "\n",
    "# get sample names that need to be imputed, in our case all samples in original dir\n",
    "base_dir = Path(CWD) / \"data\" / \"gtex_subset\"\n",
    "sample_names = sorted(d.name for d in base_dir.iterdir() if d.is_dir())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd7967b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Aruna inference...\n",
      "Loading Testing Data...\n",
      "Getting Patch Data for:\n",
      "Dataset: gtex\n",
      "Chr(s): chr21\n",
      "Patch Type: mpatch\n",
      "NR: rrbs_sim\n",
      "#Samples: 16\n",
      "Current chromosome:  chr21\n",
      "Looking for Ground-Truth Patchified FM files...\n",
      "/home/js228/ARUNA/data/gtex/patch_centric/numCpg128/true/FractionalMethylation/chr21_patches.fm.pkl\n",
      "Patchified data found at: /home/js228/ARUNA/data/gtex/patch_centric/numCpg128/true/FractionalMethylation/chr21_patches.fm.pkl\n",
      "Curent chromosome:  chr21\n",
      "Looking for Patchified Noise-Simulated FractionalMethylation and Coverage files...\n",
      "All files exist!\n",
      "FM at: /home/js228/ARUNA/data/gtex/patch_centric/numCpg128/rrbs_sim/FractionalMethylation/chr21_patches.mask.fm.pkl\n",
      "MASK at: /home/js228/ARUNA/data/gtex/patch_centric/numCpg128/rrbs_sim/SimulatedMask/chr21_patches.mask.pkl\n",
      "Loaded Validation Data with 16 samples and 288960 patches.\n",
      "#Batches in Validation (batch_dim=256): 1129\n",
      "Current chromosome:  chr21\n",
      "#CpG stats under eval: 462299.0 +/- 0.0\n"
     ]
    }
   ],
   "source": [
    "_, preds_map, evalMask_map = run_mslice_inference(test_data, model_fpath, \n",
    "                                                  config_fpath, sample_names, \n",
    "                                                  test_regime, cpgMask_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "de48419a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: /home/js228/ARUNA/results/pred_betas/chr21.csv\n"
     ]
    }
   ],
   "source": [
    "# save predictions\n",
    "save_aruna_preds(preds_map = preds_map, \n",
    "                 canonical_index = canonical_index, \n",
    "                 out_dir = os.path.join(res_dir, \"pred_betas\"),\n",
    "                 chrom = chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "46a624bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: /home/js228/ARUNA/results/eval_mask/chr21.csv\n"
     ]
    }
   ],
   "source": [
    "# save the evaluation mask\n",
    "# for simulation studies: evalMask = True for CpGs that are observed in ground truth but simulated missing\n",
    "# for true rrbs: evalMask = True for Cp Gs that are observed in matched WGBS and missing in the ground truth RRBS\n",
    "save_aruna_preds(preds_map = evalMask_map, \n",
    "                 canonical_index = canonical_index, \n",
    "                 out_dir = os.path.join(res_dir, \"eval_mask\"),\n",
    "                 chrom = chrom)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "XAI",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
